src/main/java/com/google/devtools/build/lib/analysis/starlark/FunctionTransitionUtil.java

```diff
@@ -407,36 +409,42 @@ private static void updateOutputDirectoryNameFragment(
     CoreOptions buildConfigOptions = toOptions.get(CoreOptions.class);
     Set<String> updatedAffectedByStarlarkTransition =
         new TreeSet<>(buildConfigOptions.affectedByStarlarkTransition);
-    // Add newly changed native options to overall list of changed native options
+    // Add newly changed options to overall list of changed options
     for (String option : changedOptions) {
-      if (option.startsWith(COMMAND_LINE_OPTION_PREFIX)) {
-        updatedAffectedByStarlarkTransition.add(
-            option.substring(COMMAND_LINE_OPTION_PREFIX.length()));
-      }
+      updatedAffectedByStarlarkTransition.add(option);
     }
     buildConfigOptions.affectedByStarlarkTransition =
         ImmutableList.sortedCopyOf(updatedAffectedByStarlarkTransition);
-
-    // hash all relevant native option values;
     TreeMap<String, Object> toHash = new TreeMap<>();
-    for (String nativeOption : updatedAffectedByStarlarkTransition) {
-      Object value;
-      try {
-        value =
-            optionInfoMap
-                .get(nativeOption)
-                .getDefinition()
-                .getField()
-                .get(toOptions.get(optionInfoMap.get(nativeOption).getOptionClass()));
-      } catch (IllegalAccessException e) {
-        throw new RuntimeException(
-            "IllegalAccess for option " + nativeOption + ": " + e.getMessage());
+
+    // Hash all affected native options.
+    for (String optionName : buildConfigOptions.affectedByStarlarkTransition) {
+      if (optionName.startsWith(COMMAND_LINE_OPTION_PREFIX)) {
+        String nativeOptionName = optionName.substring(COMMAND_LINE_OPTION_PREFIX.length());
+        Object value;
+        try {
+          value =
+                  optionInfoMap
+                          .get(nativeOptionName)
+                          .getDefinition()
+                          .getField()
+                          .get(toOptions.get(optionInfoMap.get(nativeOptionName).getOptionClass()));
+        } catch (IllegalAccessException e) {
+          throw new RuntimeException(
+                  "IllegalAccess for option " + nativeOptionName + ": " + e.getMessage());
+        }
+        toHash.put(optionName, value);
+      }
+    }
+
+    // Hash all affected starlark options.
+    for (Map.Entry<Label, Object> opt : toOptions.getStarlarkOptions().entrySet()) {
```

Comment:
Since we already have a for loop with an if condition above, what would you think about merging this into that?

i.e. `for () { if (native option) {} else { <this code>}} `, then with a few simple inversions of which maps are being looked up?

Reply:
The reason was to avoid invoking Label.parseAbsoluteUnchecked(...) frequently within a loop, and instead convert in the other direction from Label to String.

But I agree that an else branch, with something like:

```
     } else {
        Object value = toOptions.getStarlarkOptions().get(Label.parseAbsoluteUnchecked(optionName));
        if (value != null) {
          toHash.put(optionName, value);
        }
      }
```

could be more readable. What do you prefer?

Reply:
I'd opt for readability unless we have data justifying more complicated forms. Especially because this code is so subtle and hard-enough to properly interpret.

Reply:
OK! Resolved in 104f1dd.